Data/Primitive/Array.hs

```diff
@@ -65,12 +66,40 @@ import qualified Text.ParserCombinators.ReadPrec as RdPrc
 import Text.ParserCombinators.ReadP
 
 import Data.Functor.Classes (Eq1(..), Ord1(..), Show1(..), Read1(..))
+import Language.Haskell.TH.Syntax (Lift (..))
 
 -- | Boxed arrays.
 data Array a = Array
   { array# :: Array# a }
   deriving ( Typeable )
 
+instance Lift a => Lift (Array a) where
+#if MIN_VERSION_template_haskell(2,16,0)
+  liftTyped ary
+    | len == 0 = [|| Array (emptyArray# (##)) ||]
+    | len == 1 = [|| pure frst ||]
+    | otherwise = [|| unsafeArrayFromListN' len lst ||]
+#else
+  lift ary = [| unsafeArrayFromListN' len lst |]
+#endif
+    where
+      len = length ary
+      lst = toList ary
+      frst = ary `indexArray` 0
+
+-- | Strictly create an array from a list of a known length. If the length
+-- of the list does not match the given length, this makes demons fly
+-- out of your nose. We use it in the 'Lift' instance. If you edit the
+-- splice and break it, you get to keep both pieces.
+unsafeArrayFromListN' :: Int -> [a] -> Array a
```

Comment:
What would happen if we wrote this as a fold? Would GHC be able to make good decisions about when to build it as a list and convert it to an array vs. building it directly as an array (poke poke poke)?

Reply:
I can check this tomorrow.

Reply:
If I make the fold `INLINABLE`, it doesn't actually seem to do anything. If I make it `INLINE`, I get pokes up to around 32 elements, which suggests some judgement has been used. Unfortunately, going bigger seems to lead to some useless code duplication. So we might be better off just choosing our own cutoff, whatever that should be (currently set to 1, essentially). Another reasonable option would be to just leave it as I have it right now and push GHC for array literals or something.